sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs

```diff
@@ -0,0 +1,144 @@
+            }
+            else
+            {
+                // We need a multiple of 512 to flush.
+                if (_buffer.Length % Constants.Blob.Page.PageSize != 0)
+                {
+                    int bytesToWrite = (int)(Constants.Blob.Page.PageSize - _buffer.Length % Constants.Blob.Page.PageSize);
+                    await WriteToBufferInternal(buffer, offset, bytesToWrite, async, cancellationToken).ConfigureAwait(false);
+                    remaining -= bytesToWrite;
+                    offset += bytesToWrite;
+                }
+
+                // Flush the buffer.
+                await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+                while (remaining > 0)
+                {
+                    await WriteToBufferInternal(
+                        buffer,
+                        offset,
+                        (int)Math.Min(remaining, _bufferSize),
+                        async,
+                        cancellationToken).ConfigureAwait(false);
+
+                    // Remaining bytes won't fit in buffer.
+                    if (remaining > _bufferSize)
+                    {
+                        await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+                        remaining -= (int)_bufferSize;
+                        offset += (int)_bufferSize;
+                    }
+
+                    // Remaining bytes will fit in buffer.
+                    else
+                    {
+                        remaining = 0;
+                    }
+                }
+            }
+            _position += count;
+        }
+
+        protected override async Task AppendInternal(bool async, CancellationToken cancellationToken)
+        {
+            if (_buffer.Length > 0)
+            {
+                _buffer.Position = 0;
+
+                Response<PageInfo> response = await _pageBlobClient.UploadPagesInternal(
+                    content: _buffer,
+                    offset: _writeIndex,
+                    transactionalContentHash: default,
+                    conditions: _conditions,
+                    progressHandler: _progressHandler,
+                    async: async,
+                    cancellationToken: cancellationToken)
+                    .ConfigureAwait(false);
+
+                _conditions.IfMatch = response.Value.ETag;
+
+                _writeIndex += _buffer.Length;
+                _buffer.Clear();
+            }
+        }
+
+        protected override async Task FlushInternal(bool async, CancellationToken cancellationToken)
+            => await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+        protected override void ValidateBufferSize(long bufferSize)
+        {
+            if (bufferSize < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be >= 1");
```

Comment:
Nit - make the message format match the base class

Reply:
Changed to `throw new ArgumentOutOfRangeException(nameof(bufferSize), "Must be greater than or equal to 1");`.  Note that we are checking the buffer size here, not `offset` or `count`.